chapters/grid-search.qmd

```diff
@@ -384,7 +383,497 @@ A simple workaround for creating a space-filling design is to repeat the unique
+
+```{r}
+#| label: fig-racing-initial
+#| echo: false
+#| out-width: 70%
+#| fig-align: center
+#| fig-width: 6
+#| fig-height: 4.25
+#| fig-cap: The first interim analysis in racing using a larger space-filling design for the boosted tree model.
+
+brier_param %>% 
+  ggplot(aes(y = candidate)) + 
+  geom_point(aes(x = estimate), cex = 1 / 2) +
+  geom_vline(xintercept = 0, lty = 2) +
+  geom_errorbar(aes(xmin = conf.low, xmax = estimate, col = Decision), 
+                alpha = 1 / 3) +
+  labs(x = "Drop in Brier Score", y = "Candidate Rank")
+```
+
+In the end, the racing process eliminated all but `r num_remaining` candidates. The process eliminated candidates at iterations `r loss_text[[1]]`. In all, `r num_fit` models were fit out of the possible `r format(nrow(large_sfd) * nrow(sim_rs), big.mark = ",")` (`r round(num_fit / (nrow(large_sfd) * nrow(sim_rs)) * 100, 1)`%). `r best_text`
+
+The model in @eq-perf-mod-racing allows us to estimate the within-resample correlation coefficient. This estimates how similar the performance metric values are to one another relative to values between resamples. The correlation estimate is `r signif(within_corr, 2)`, indicating a moderate similarity of metrics within a resample. 
+
+In summary, racing can be an effective tool for screening a large number of models while using comprehensive grids. 
+
+
+## Optimization Bias and Nested Resampling {#sec-nested-resampling}
+
+```{r}
+#| label: 1d-boost
+#| include: false
+# Objects from nested_res.RData
+nest_grid_brier <- collect_metrics(nest_grid_res)
+grid_brier <- show_best(grid_res, metric = "brier_class", n = 1)
+opt_per_resample <- 
+  collect_metrics(grid_res, summarize = FALSE) %>%
+  slice_min(.estimate, by = id) 
+multiples <- 
+  opt_per_resample %>% 
+  count(learn_rate) %>% 
+  filter(n > 1) %>% 
+  pluck("learn_rate") %>% 
+  log10() %>% 
+  signif(., digits = 3)
+lr_lab <- cli::pluralize("There {?was/were} {length(multiples)} learning rate{?s} selected more than once: ") 
+lr_lab <- paste0(lr_lab, knitr::combine_words(paste0("10<sup>", multiples, "</sup>")), ".")
+```
+
+In the last section, we found that the numerically best candidate was `r best_large`. The corresponding resampling estimate of the Brier score was `r signif(best_grid$mean, 3)` with a standard error of `r signif(best_grid$std_err, 3)`. If someone were to ask us how well the optimal boosted tree performs, we would probably give them these performance estimates. 
+
+However, there may be an issue in doing so. We are using grid search and resampling to find the best combination of parameters _and_ to estimate the corresponding performance. the problem is that we don't know how accurate or precise the estimate of the _optimal_ candidate may be. In some situations, this dual use of the model tuning process can introduce **optimization bias** where, to some degree, our performance statistics are optimistic. 
```

Comment:
```suggestion
However, there is an issue in doing so that may affect our decision for selecting the optimal tuning parameters. We are using grid search and resampling to find the best combination of parameters _and_ to estimate the corresponding performance. The problem is that we don't know how accurate or precise the estimate of the _optimal_ candidate may be. In some situations, this dual use of the model tuning process can introduce **optimization bias** where, to some degree, our performance statistics are optimistic. 
```